EVA-2290 - Refactor the reverse_strand.py

tcezard opened the pull request:
Refactor and add unit tests

tcezard committed:
Refactor the reverse_strand.py script to allow testing of filtering features

sundarvenkata-EBI commented:
I understand that actions have been set up as part of this PR and will kick into action after merge. But would you be able to run the github action **manually**  just for this PR?

sundarvenkata-EBI commented on variant_remapping_tools/tests/test_reverse_strand.py:
```diff
@@ -0,0 +1,81 @@
+from unittest.mock import Mock
+
+import pysam
+from pysam.libcalignedsegment import AlignedSegment
```
This does not look like it is being used...

tcezard replied:
removed

sundarvenkata-EBI commented on variant_remapping_tools/reverse_strand.py:
```diff
@@ -72,43 +20,116 @@
+
+def process_bam_file(bam_file_path, output_file, old_ref_allele_output_file, flank_length, score_perc, diff_AS_XS_perc):
+    bamfile = pysam.AlignmentFile(bam_file_path, 'rb')
+    outfile = open(output_file, 'w')
```
Suggest using the [with expression](https://docs.python.org/2.5/whatsnew/pep-343.html) to close write file handles in case of errors.

tcezard replied:
Didn't want to change anything but I guess that won't change the logic.
Done

tcezard committed:
address comments from review

tcezard commented:
>I understand that actions have been set up as part of this PR and will kick into action after merge. But would you be able to run the github action manually just for this PR?

Turns out that the github action was being run nut failed because of the yaml syntax. This should now be running properly.

tskir commented on .github/workflows/variant_remapping.yml:
```diff
@@ -0,0 +1,40 @@
+        python-version: ${{ matrix.python-version }}
+    - name: Install dependencies
+      run: |
+        # Install Nextflow
+        mkdir /tmp/nextflow && cd /tmp/nextflow
+        wget -qO- get.nextflow.io | bash
+        echo "/tmp/nextflow" >> $GITHUB_PATH
+        cd -
```
Nextflow is actually [available through Bioconda](https://bioconda.github.io/recipes/nextflow/README.html), so seems like this bit could be replaced by just adding it as a dependency in `conda.yml`. An additional rationale for that is that currently this step is very noisy in the logs and adds hundreds of lines like this:
![image](https://user-images.githubusercontent.com/10669118/107019824-0b787400-679a-11eb-9a13-feb6935d534a.png)

tcezard replied:
Made the the conda install quiet with `-q` and move the nextflow install to conda as well. 
Good call

tskir commented on .github/workflows/variant_remapping.yml:
```diff
@@ -0,0 +1,40 @@
+        # Install dependencies
+        # $CONDA is an environment variable pointing to the root of the miniconda directory
+        $CONDA/bin/conda env update --file conda.yml --name base
+        $CONDA/bin/conda run pip install -r requirements.txt
```
```suggestion
        $CONDA/bin/conda run pip -q install -r requirements.txt
```

This will still show errors should they occur, but will reduce the log output overall

tcezard replied:
Done

tskir commented on requirements.txt:
```diff
@@ -1,2 +1,3 @@
-biopython==1.77
\ No newline at end of file
+biopython==1.77
+pytest
```
```suggestion
pytest==6.2.2
```

Since the other two dependencies are pinned, suggest pinning this one as well

tskir commented on variant_remapping_tools/reverse_strand.py:
```diff
@@ -72,43 +20,114 @@
+    # Getting AS and XS:
+    AS = read.get_tag("AS")
+    try:
+        XS = read.get_tag("XS")  # Some variants don't have secondary alignments, which throws an error
```
This is a super fine level nitpick, but currently there's a mixture of single and double quotes for string literals—maybe now's the time to fix it, before the module grows and it becomes more cumbersome

tskir replied:
(I think single quotes are better but that's a personal choice. PIP 8 just recommends the choice is consistent throughout the codebase.)

tcezard replied:
That's true and not difficult to change. This is now done.

tskir commented on variant_remapping_tools/reverse_strand.py:
```diff
@@ -72,43 +20,114 @@
+def process_bam_file(bam_file_path, output_file, old_ref_allele_output_file, flank_length, score_perc, diff_AS_XS_perc):
+    bamfile = pysam.AlignmentFile(bam_file_path, 'rb')
+
+    # Calculate the score cutoff based on flanking seq length:
```
```suggestion
    # Calculate the score cutoff based on flanking seq length
```

tskir commented on variant_remapping_tools/tests/test_reverse_strand.py:
```diff
@@ -0,0 +1,80 @@
+        self.diff_cutoff = .06
+
+    def test_calculate_variant_position(self):
+        # 101M: Perfect alignment
```
I think if you convert comments to one-line docstrings, pytest will be able to automatically use them for test "descriptions" and error messages in case something fails

tcezard replied:
I did not know that. That a good trick.

tskir commented on variant_remapping_tools/tests/test_reverse_strand.py:
```diff
@@ -0,0 +1,80 @@
+        read = self._mock_read(cigartuples=((pysam.CMATCH, 51), (pysam.CDEL, 2), (pysam.CMATCH, 50)))
+        assert is_read_valid(read, self.counter, 50, self.score_cutoff, self.diff_cutoff) is False
+        assert self.counter == Counter({'total': 1, 'context_bad': 1})
+
+
```
```suggestion

```

tskir reviewed (approved):
Just some minor nitpicks. Love the tests BTW, succinct and to the point!

tcezard committed:
address comments from review:
 - cleanup quotes
 - use conda for netflow install
 - add docstrings in place of comments

tcezard committed:
use quiet mode for conda install

tcezard committed:
Apply suggestions from code review

Co-authored-by: Kirill Tsukanov <tskir@users.noreply.github.com>
